gk/main.c

```diff
@@ -552,6 +552,71 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_lpm *ltbl = arg;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4
+				&& memcmp(&ltbl->tunnels[i].flow.f.v4.dst,
+				map->ip_be, sizeof(struct in_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
```

Comment:
I thought about this problem a lot when writing the LLS block, before we decided to push the responsibility of what to do with an update to a callback function. It's tricky because these writes are extremely rare, but the reads are extremely common. Here's what the DPDK programmer's guide says about locks and atomic operations:

> Atomic operations imply a lock prefix before the instruction, causing the processor’s LOCK# signal to be asserted during execution of the following instruction. This has a big impact on performance in a multicore environment.
> 
> Performance can be improved by avoiding lock mechanisms in the data plane. It can often be replaced by other solutions like per-lcore variables. Also, some locking techniques are more efficient than others. For instance, the Read-Copy-Update (RCU) algorithm can frequently replace simple rwlocks.

So it's tempting to use an atomic variable or a rwlock here for simplicity, but this warning about locks in the data plane makes me unsure.